### PR TITLE
Refactor File Operations to a Module

### PR DESCRIPTION
**Pull request by nvisal1**
***Purpose of this PR***
* Move single file download logic to the `FileManager` module
* Expose functions for deleting files and folders in the `FileManager` module
* Refactor `FileManager` module to follow the Service Module pattern
* Remove legacy file upload code
* Remove the `URL` property from Learning Object PDF -> `breaking change`

### CODE REVIEW

File: src/FileManager/adapters/FileManagerAdapter.ts [hunk 1]
```diff
@@ -1,15 +1,11 @@
-import { FileManager } from '../shared/interfaces/interfaces';
-import { uploadFile } from './FileInteractor';
-import { FileUpload } from '../shared/interfaces/FileManager';
-export class FileManagerAdapter {
+import { uploadFile } from '../Interactor';
+import { FileUpload } from '../typings/file-manager';
+import { FileManagerAdapter as FileManagerAdapterInterface } from '../interfaces/FileManagerAdapter';
+export class FileManagerAdapter implements FileManagerAdapterInterface {
```
Comment (gshaw1997) [inline src/FileManager/adapters/FileManagerAdapter.ts]: If using the `ServiceModule` could we export the functions we want to expose from this module like done [here](https://github.com/Cyber4All/learning-object-service/pull/384#discussion_r297806866), instead for consistency. I know this isn't documented anywhere yet, but will do so when going through and writing up comprehensive docs for using the pattern.

File: src/FileManager/adapters/FileManagerAdapterStub.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+
```
Comment (gshaw1997) [inline src/FileManager/adapters/FileManagerAdapterStub.ts]: What's the purpose of this stub? Shouldn't the dependent of this dependency provide this mock for tests?

File: src/FileManager/typings/file-manager.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
+import { CompletedPartList, CompletedPart } from 'aws-sdk/clients/s3';
+
+// Export aliased types for ease of update in the case that AWS S3 is no longer the driver used.
+export type CompletedPart = CompletedPart;
+export type CompletedPartList = CompletedPartList;
+
+export interface FileUpload {
```
Comment (gshaw1997) [inline src/FileManager/typings/file-manager.ts]: Is this still needed?

File: src/FileManager/typings/file-manager.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    path: string;
+    // FIXME: This should define the specific types it can take
+    data: any;
+  }
+
+export interface MultipartFileUpload extends FileUpload {
```
Comment (gshaw1997) [inline src/FileManager/typings/file-manager.ts]: Is this still needed?

File: src/FileManager/typings/file-manager.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    partNumber: number;
+    uploadId: string;
+}
+
+export interface MultipartFileUploadStatus {
```
Comment (gshaw1997) [inline src/FileManager/typings/file-manager.ts]: Is this still needed?

File: src/FileManager/typings/file-manager.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    _id: string;
+    path: string;
+    completedParts: CompletedPartList;
+    createdAt: string;
+}
+export interface MultipartFileUploadStatusUpdates {
```
Comment (gshaw1997) [inline src/FileManager/typings/file-manager.ts]: Is this still needed?

File: src/FileManager/typings/file-manager.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    partsUploaded: number;
+    bytesUploaded: number;
+}
+
+export interface MultipartUploadData {
```
Comment (gshaw1997) [inline src/FileManager/typings/file-manager.ts]: Is this still needed?

File: src/FileManager/typings/file-manager.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    uploadId?: string;
+    url?: string;
+    completedPart?: CompletedPart;
+}
+
+export interface DZFileMetadata {
```
Comment (gshaw1997) [inline src/FileManager/typings/file-manager.ts]: Is this still needed?

File: src/FileManager/typings/file-manager.ts [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    dzchunkindex?: number;
+    dztotalfilesize?: number;
+    dzchunksize?: number;
+    size?: number;
+    dztotalchunkcount?: number;
+    dzchunkbyteoffset?: number;
+    fullPath: string;
+}
+
+export interface DZFile extends DZFileMetadata {
```
Comment (gshaw1997) [inline src/FileManager/typings/file-manager.ts]: Is this still needed?

File: src/FileMetadata/adapters/FileMetadataAdapter.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
+import { LearningObjectFile, FileMetadataFilter } from '../typings';
```
Comment (gshaw1997) [inline src/FileMetadata/adapters/FileMetadataAdapter.ts]: An adapter isn't needed here. The functions are already exposed via the module

File: src/LearningObjects/PDFKitDriver.ts [hunk 11]
```diff
@@ -1,9 +1,9 @@
 import * as PDFKit from 'pdfkit';
 import { LEARNING_OBJECT_ROUTES } from '../shared/routes';
 import { FileManager } from '../shared/interfaces/interfaces';
-import { FileUpload } from '../shared/interfaces/FileManager';
 import * as striptags from 'striptags';
 import { LearningObject, LearningOutcome } from '../shared/entity';
+import { FileUpload } from '../FileManager/typings/file-manager';
```
Comment (gshaw1997) [inline src/LearningObjects/PDFKitDriver.ts]: What is this used for?

File: src/LearningObjects/LearningObjectRouteHandler.spec.ts [hunk 12]
```diff
@@ -48,7 +48,7 @@ describe('LearningObjectRouteHandler', () => {
         dataStore = await MongoDriver.build(global['__MONGO_URI__']);
         HierarchyAdapter.open(dataStore);
         fileManager = new MockS3Driver();
-        FileManagerAdapter.open(fileManager);
+        FileManagerAdapter.open();
```
Comment (gshaw1997) [inline src/LearningObjects/LearningObjectRouteHandler.spec.ts]: If using the `ServiceModule` we'll want to initialize the module, but shouldn't this dependency be stubbed/mocked?

File: src/app.ts [hunk 13]
```diff
@@ -85,7 +85,7 @@ function initModules() {
   HierarchyAdapter.open(dataStore);
   LearningObjectAdapter.open(dataStore, fileManager, library);
   LearningObjectSearch.initialize();
-  FileManagerAdapter.open(fileManager);
+  FileManagerAdapter.open();
```
Comment (gshaw1997) [inline src/app.ts]: If using the `ServiceModule` we'll want to initialize here.

File: src/drivers/MongoDriver.ts [hunk 14]
```diff
@@ -1933,6 +1821,39 @@ export class MongoDriver implements DataStore {
     return query;
   }
 
+  async findSingleFile(params: {
```
Comment (gshaw1997) [inline src/drivers/MongoDriver.ts]: Is this still needed in the MongoDriver?

File: src/shared/interfaces/interfaces.ts [hunk 15]
```diff
@@ -1,6 +1,6 @@
 
 import { DataStore } from './DataStore';
-import { FileManager } from './FileManager';
+import { FileManager } from '../../FileManager/interfaces/FileManager';
```
Comment (gshaw1997) [inline src/shared/interfaces/interfaces.ts]: Do we need this interface exposed at the global service level?

File: src/tests/mock-drivers/MockS3Driver.ts [hunk 16]
```diff
@@ -1,7 +1,7 @@
 import { FileManager } from '../../shared/interfaces/interfaces';
-import { FileUpload, CompletedPartList } from '../../shared/interfaces/FileManager';
 import { CompletedPart } from 'aws-sdk/clients/s3';
 import { Readable } from 'stream';
+import { FileUpload, CompletedPartList } from '../../FileManager/typings/file-manager';
```
Comment (gshaw1997) [inline src/tests/mock-drivers/MockS3Driver.ts]: Is this still needed?

File: src/shared/interfaces/DataStore.ts [hunk 17]
```diff
@@ -206,12 +191,10 @@ export interface DataStore
     fileId: string;
     description: string;
   }): Promise<LearningObject.Material.File>;
-
-  // File Uploads
-  updateMultipartUploadStatus(params: {
-    id: string;
-    completedPart: CompletedPart;
-  }): Promise<void>;
+  findSingleFile(params: {
```
[the remark this answers is not in the document]
Comment (nvisal1) [reply src/shared/interfaces/DataStore.ts]: I decided to keep this function because it is used by the `updateFileDescription` function in the `MongoDriver` and the `removeFile` function in `LearningObjects/LearningObjectInteractor`.

Can you confirm if these functions are deprecated?

File: src/drivers/drivers.ts [hunk 18]
```diff
@@ -3,7 +3,7 @@ import { ExpressRouteDriver } from './express/ExpressRouteDriver';
 import { ExpressAuthRouteDriver } from './express/ExpressAuthRouteDriver';
 import { ExpressAdminRouteDriver } from './express/ExpressAdminRouteDriver';
 import { MongoDriver } from './MongoDriver';
-import { S3Driver } from '../FileManager/S3Driver';
+import { S3Driver } from '../FileManager/drivers/FileManagerDriver/S3Driver';
```
[the remark this answers is not in the document]
Comment (nvisal1) [reply src/drivers/drivers.ts]: It is being used in the `LearningObjects` module in the `updateReadme` function.

File: src/FileManager/Interactor.ts [hunk 19]
```diff
@@ -54,7 +104,7 @@ export async function downloadSingleFile(params: {
   }
 
   if (!params.filter || params.filter === 'released') {
-    fileMetaData = await FileMetadata.geFileMetadata({
+    fileMetaData = await FileMetadata.getFileMetadata({
```
Comment (gshaw1997) [inline src/FileManager/Interactor.ts]: We should swap out direct call to the module with a call to an interface that uses this module, like a module gateway

File: src/FileManager/Interactor.ts [hunk 20]
```diff
@@ -70,7 +120,7 @@ export async function downloadSingleFile(params: {
 
     // To maintain existing functionality, service token is elevated to author to fetch unreleased file meta
     serviceToken.username = learningObject.author.username;
-    fileMetaData = await FileMetadata.geFileMetadata({
+    fileMetaData = await FileMetadata.getFileMetadata({
```
Comment (gshaw1997) [inline src/FileManager/Interactor.ts]: We should swap out direct call to the module with a call to an interface that uses this module, like a module gateway

File: src/FileManager/Interactor.ts [hunk 21]
```diff
@@ -1,11 +1,18 @@
+import { FileManagerModule as Module } from '.';
 import { DataStore } from '../shared/interfaces/DataStore';
```
[the remark this answers is not in the document]
Comment (nvisal1) [reply src/FileManager/Interactor.ts]: It is used in the `Drivers` namespace to resolve module dependencies.

File: src/FileMetadata/drivers/LearningObjectGateway/ModuleLearningObjectGateway.ts [hunk 22]
```diff
@@ -5,6 +5,8 @@ import {
   LearningObjectFile,
 } from '../../typings';
 import { LearningObjectAdapter } from '../../../LearningObjects/LearningObjectAdapter';
+import { UserToken } from '../../../shared/types';
```
Comment (gshaw1997) [inline src/FileMetadata/drivers/LearningObjectGateway/ModuleLearningObjectGateway.ts]: Can we move this to a typings internal to the module?

File: src/FileMetadata/drivers/LearningObjectGateway/ModuleLearningObjectGateway.ts [hunk 23]
```diff
@@ -5,6 +5,8 @@ import {
...
+import { LearningObject } from '../../../shared/entity';
```
Comment (gshaw1997) [inline src/FileMetadata/drivers/LearningObjectGateway/ModuleLearningObjectGateway.ts]: Can we move this to a typings internal to the module?

File: src/FileManager/adapters/FileManagerAdapter.ts (hunk 1 above)
Comment (nvisal1) [reply src/FileManager/adapters/FileManagerAdapter.ts]: Yeah, I totally missed that thread.

File: src/shared/interfaces/DataStore.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (gshaw1997) [reply src/shared/interfaces/DataStore.ts]: Ah I see, yes they are now deprecated :ok_hand:

File: src/drivers/drivers.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (gshaw1997) [reply src/drivers/drivers.ts]: Could we swap this for a FileGateway that the `updateReadme` function could use?

File: src/FileManager/Interactor.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (gshaw1997) [reply src/FileManager/Interactor.ts]: Ah I see, idk how I missed that haha. Thanks for clarifying.

File: src/FileManager/adapters/FileManagerAdapterStub.ts (hunk 2 above)
Comment (nvisal1) [reply src/FileManager/adapters/FileManagerAdapterStub.ts]: I'm not certain. I believe it was authored by @DonSeannelly.

File: src/FileManager/typings/file-manager.ts (hunk 3 above)
Comment (nvisal1) [reply src/FileManager/typings/file-manager.ts]: This is still used in many places.

File: src/LearningObjects/PDFKitDriver.ts (hunk 11 above)
Comment (nvisal1) [reply src/LearningObjects/PDFKitDriver.ts]: This function is used by the `function addEventListeners` function.

File: src/LearningObjects/LearningObjectRouteHandler.spec.ts (hunk 12 above)
Comment (nvisal1) [reply src/LearningObjects/LearningObjectRouteHandler.spec.ts]: The FileManagerAdapter was deleted in favor of exposing static functions at the module index.

File: src/app.ts (hunk 13 above)
Comment (nvisal1) [reply src/app.ts]: The FileManagerAdapter was deleted in favor of exposing static functions at the module index.

File: src/drivers/MongoDriver.ts (hunk 14 above)
Comment (nvisal1) [reply src/drivers/MongoDriver.ts]: I decided to keep this function because it is used by the updateFileDescription function in the MongoDriver and the removeFile function in LearningObjects/LearningObjectInteractor.

Can you confirm if these functions are deprecated?

File: src/tests/mock-drivers/MockS3Driver.ts (hunk 16 above)
Comment (nvisal1) [reply src/tests/mock-drivers/MockS3Driver.ts]: FileUpload will still be needed. I think this that file should be moved into the `FileManager` module. Thoughts?

File: src/FileMetadata/drivers/LearningObjectGateway/ModuleLearningObjectGateway.ts (hunk 22 above)
Comment (nvisal1) [reply src/FileMetadata/drivers/LearningObjectGateway/ModuleLearningObjectGateway.ts]: Should `UserToken` be reimplemented in every module that uses it?

File: src/FileMetadata/drivers/LearningObjectGateway/ModuleLearningObjectGateway.ts (hunk 23 above)
Comment (nvisal1) [reply src/FileMetadata/drivers/LearningObjectGateway/ModuleLearningObjectGateway.ts]: Should `Entity` types be redefined in every module that uses them?

File: src/drivers/MongoDriver.ts (hunk 14 above)
Comment (gshaw1997) [reply src/drivers/MongoDriver.ts]: They are indeed deprecated

File: src/tests/mock-drivers/MockS3Driver.ts (hunk 16 above)
Comment (gshaw1997) [reply src/tests/mock-drivers/MockS3Driver.ts]: I would agree, we should move all file ops to the `FileManager`

File: src/shared/interfaces/interfaces.ts (hunk 15 above)
Comment (nvisal1) [reply src/shared/interfaces/interfaces.ts]: Mock S3 driver should be a stub gateway that lives in the Learning Object module

File: src/FileManager/Interactor.ts [hunk 24]
```diff
@@ -0,0 +1,261 @@
...
+  authorUsername: string;
+  learningObjectId: string;
+  path: string;
+}): Promise<void> {
+  await Drivers.fileManager().deleteFolder({
+    authorUsername,
+    learningObjectId,
+    path,
+  });
+}
```
Comment (DonSeannelly) [inline src/FileManager/Interactor.ts]: A few of these interactor functions just proxy requests to the fileManager. What was the reason for this, as compared to giving other modules access to inject this driver?

File: src/FileManager/Interactor.ts [hunk 25]
```diff
@@ -0,0 +1,261 @@
...
+
+/**
+ * Sends a file back to the caller as a readable stream.
+ *
```
Comment (DonSeannelly) [inline src/FileManager/Interactor.ts]: There's quite a bit of complexity here in this function. We should update the doc to note the nuances.

File: src/FileManager/drivers/FileMetadataGateway/FileMetadataModuleGateway.ts [hunk 26]
```diff
@@ -0,0 +1,24 @@
...
+     */
+    getFileMetadata(params: {
+        requester: UserToken;
+        learningObjectId: string;
+        id: string;
+        filter: FileMetadataFilter;
+    }): Promise<LearningObject.Material.File> {
+        return FileMetadataModule.getFileMetadata(params);
+    }
+}
```
Comment (DonSeannelly) [inline src/FileManager/drivers/FileMetadataGateway/FileMetadataModuleGateway.ts]: What is the driving factor behind the `ModuleGateway` proxy pattern?

Just sort of thinking here, but why not just have the module declare an interface that it expects, and then drop the `FileMetadataModule` in?

File: src/LearningObjects/Publishing/Bundler/Interactor.ts [hunk 27]
```diff
@@ -96,21 +95,18 @@ function addCCLicense(prefix: string = ''): BundleData {
...
-  uri,
   name,
   prefix = '',
 }: {
-  uri: string;
   name: string;
   prefix?: string;
 }): BundleData {
-  return { name, prefix, uri };
+  return { name, prefix, uri: null };
```
Comment (DonSeannelly) [inline src/LearningObjects/Publishing/Bundler/Interactor.ts]: Is this permanent, or should we have a FIXME/TODO?

File: src/LearningObjects/Publishing/interactor.ts [hunk 28]
```diff
@@ -25,22 +29,32 @@ export interface PublishingDataStore {
...
-    }
-    await dataStore.addToReleased(releasableObject);
-    await sendEmail(releasableObject, userToken, releaseEmailGateway);
+  if (!requesterIsAdminOrEditor(userToken)) {
+    throw new ResourceError(
+      `${
+        userToken.username
+      } does not have access to release this Learning Object`,
+      ResourceErrorReason.INVALID_ACCESS,
+    );
```
Comment (DonSeannelly) [inline src/LearningObjects/Publishing/interactor.ts]: @gshaw1997 Can we update local prettier config to match the repo's tslint rules (or vice versa) so that we can avoid all of these diffs

File: src/LearningObjects/Publishing/interactor.ts [hunk 29]
```diff
@@ -52,25 +66,22 @@ export async function releaseLearningObject({ userToken, dataStore, releasableOb
...
  * @param userToken the user who has requested to publish a Learning Object
  */
-async function createPublishingArtifacts(releasableObject: LearningObject, userToken: UserToken) {
-    const storagePrefix = `${releasableObject.author.username}/${releasableObject.id}`;
-    FileManagerAdapter.getInstance().uploadFile({
-        file: {
-            path: `${storagePrefix}/meta.json`,
-            data: JSON.stringify(releasableObject.toPlainObject()),
-        },
-    });
```
Comment (DonSeannelly) [inline src/LearningObjects/Publishing/interactor.ts]: @gshaw1997 It looks like we removed creation of the `meta.json`. Was this an accident?

File: src/drivers/MongoDriver.ts [hunk 30]
```diff
@@ -1832,64 +1687,6 @@ export class MongoDriver implements DataStore {
...
-      );
-      return this.findSingleFile({
-        learningObjectId: params.learningObjectId,
-        fileId: params.fileId,
-      });
-    } catch (e) {
-      return Promise.reject(e);
-    }
-  }
-
```
Comment (DonSeannelly) [inline src/drivers/MongoDriver.ts]: image

File: src/shared/functions/index.ts [hunk 31]
```diff
@@ -130,3 +130,30 @@ export function mapAuthorToSummary(author: Partial<User>): AuthorSummary {
     organization: author.organization,
   };
 }
+
+/**
+ * Title cases string
```
Comment (DonSeannelly) [inline src/shared/functions/index.ts]: This JSDoc description says nothing more than the function name does. Small nitpick here, but maybe something like: 
```suggestion
 * Capitalizes the first letter of each word in the provided string.
```

File: src/shared/types/file-upload.ts [hunk 32]
```diff
@@ -0,0 +1,7 @@
+import { Readable } from 'stream';
+
+export interface FileUpload {
+    path: string;
+    // FIXME: This should define the specific types it can take
+    data: Buffer|Uint8Array|Blob|string|Readable;
```
Comment (DonSeannelly) [inline src/shared/types/file-upload.ts]: It looks like this FIXME may have been resolved? I believe I put it here when `data` had a type of `any`.

File: src/FileManager/drivers/FileMetadataGateway/FileMetadataModuleGateway.ts (hunk 26 above)
Comment (gshaw1997) [reply src/FileManager/drivers/FileMetadataGateway/FileMetadataModuleGateway.ts]: It was supposed to serve as an abstraction layer, but you make a good point. If we declare the interface the interactor is expecting, we'll be able to drop in the module directly as the dependency, as long as those two interfaces are consistent. What do we do in the case where they are not consistent? Do would we instead create this `ModuleGateway` (or at that point `Adapter` may be a more appropriate name)? If so, why not just do that from the start, to maintain consistency?

File: src/LearningObjects/Publishing/Bundler/Interactor.ts (hunk 27 above)
Comment (gshaw1997) [reply src/LearningObjects/Publishing/Bundler/Interactor.ts]: This should have a FIXME, but it is fixed in this PR https://github.com/Cyber4All/learning-object-service/pull/404 since pdf urls were removed.

File: src/LearningObjects/Publishing/interactor.ts (hunk 29 above)
Comment (gshaw1997) [reply src/LearningObjects/Publishing/interactor.ts]: Yup, was definitely an accident. Good catch!

File: src/FileManager/Interactor.ts (hunk 24 above)
Comment (gshaw1997) [reply src/FileManager/Interactor.ts]: We wanted to move all business logic related to file management to a single place and have the `FileManager` module own that domain. Although some functions just proxy the `fileManager` driver, there are operations that require logic.

File: src/FileManager/drivers/FileMetadataGateway/FileMetadataModuleGateway.ts (hunk 26 above)
Comment (DonSeannelly) [reply src/FileManager/drivers/FileMetadataGateway/FileMetadataModuleGateway.ts]: Personally, I'm a fan of keeping it as simple as possible and only scaling up the patterns in use when needed. I haven't yet seen us actually need to use an Adapter in this way, so I would say if we need to in the future we can easily set that up, but in the other cases where it is a direct mapping we can keep it simpler.

File: src/LearningObjects/Publishing/interactor.ts (hunk 28 above)
Comment (gshaw1997) [reply src/LearningObjects/Publishing/interactor.ts]: Is there a good way to sync the two? Most options don't seem to have a one-to-one mapping from ts-lint config and prettier config

File: tslint.json [hunk 33]
```diff
@@ -11,7 +11,7 @@
     "interface-name": [false],
     "jsdoc-format": true,
     "label-position": true,
-    "max-line-length": [true, 180],
+    "max-line-length": [true, 80],
```
Comment (DonSeannelly) [inline tslint.json]: @gshaw1997 What’s the reasoning for choosing the number 80?

File: .prettierrc [hunk 34]
```diff
@@ -0,0 +1,14 @@
...
+  "htmlWhitespaceSensitivity": "css",
+  "jsxBracketSameLine": true,
+  "jsxSingleQuote": true,
+  "printWidth": 80,
+  "proseWrap": "preserve",
+  "quoteProps": "as-needed",
+  "semi": true,
+  "tabWidth": 2,
+  "trailingComma": "es5"
+}
```
Comment (DonSeannelly) [inline .prettierrc]: This commit might be better off going in its own PR. How do we plan to include prettier in each developers workflow? I know there’s a VS code extension and a CLI. How can we make this consistent for developers? I’m thinking we might need to update the read me with some extra steps for getting started developing in this repository

File: tslint.json (hunk 33 above)
Comment (gshaw1997) [reply tslint.json]: This decision was based on the recommendation by Prettier https://prettier.io/docs/en/options.html#print-width

File: src/FileManager/Interactor.ts (hunk 24 above)
Comment (DonSeannelly) [reply src/FileManager/Interactor.ts]: I think that I’m okay with this for now, but the question that’s coming up in my mind is perhaps we need to identify a way to break up business logic when it’s not quite an interactor by the traditional sense
